src/functional.rs

```diff
@@ -0,0 +1,98 @@
+//! Functional programming with [`Array`]s.
+// This is modeled after `generic-array::functional`
+// see: <https://docs.rs/generic-array/1.0.0/generic_array/functional/index.html>
+
+use crate::{Array, ArraySize};
+
+impl<T, U> Array<T, U>
+where
+    U: ArraySize,
+{
+    /// Maps an `Array` to another `Array`
+    ///
+    /// If the mapping function panics, any already initialized elements in the new array will
+    /// be dropped, AND any unused elements in the source array will also be dropped.
+    #[inline(always)]
+    pub fn map<O, F>(self, f: F) -> Array<O, U>
+    where
+        F: FnMut(T) -> O,
+    {
+        Mapper {
+            inner: self.into_iter(),
+            f,
+        }
+        .collect()
+    }
+
+    /// Combines two `Array` instances and iterates through both of them, initialization a new
+    /// `Array` with the result of the zipped mapping function.
+    ///
+    /// If the mapping function panics, any already initialized elements in the new array will
+    /// be dropped, AND any unused elements in the source arrays will also be dropped.
+    #[inline(always)]
+    pub fn zip<Rhs, F, O>(self, rhs: Array<Rhs, U>, f: F) -> Array<O, U>
+    where
+        U: ArraySize,
+        F: FnMut(T, Rhs) -> O,
+    {
+        Zipper {
+            inner: self.into_iter(),
+            rhs: rhs.into_iter(),
+            f,
+        }
+        .collect()
+    }
+
+    /// Folds (or reduces) a sequence of data into a single value.
+    ///
+    /// If the fold function panics, any unused elements will be dropped.
+    #[inline(always)]
+    pub fn fold<O, F>(self, mut init: O, mut f: F) -> O
+    where
+        F: FnMut(O, T) -> O,
+    {
+        for t in self {
+            init = f(init, t)
+        }
+
+        init
+    }
```

Comment:
This feels like it could just be `self.into_iter().fold(...)` (i.e. is there really value in an inherent method over using `Iterator`?)

Reply:
This was mostly to keep the original api. Otherwise, agreed.